File: src/main/resources/docker/java/compile/CompileWorker.java
import org.apache.kafka.clients.consumer.*;
import org.apache.kafka.clients.producer.*;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.apache.kafka.common.serialization.StringSerializer;
import org.apache.commons.text.StringEscapeUtils;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.file.*;
import java.time.Duration;
import java.util.*;

public class CompileWorker {
    public static void main(String[] args) throws Exception {
        String kafkaServer = System.getenv().getOrDefault("KAFKA_SERVERS", "http://host.docker.internal:9092");
        String resultEndpoint = System.getenv().getOrDefault("RESULT_ENDPOINT", "http://host.docker.internal:8082/submit/result");

        Properties consumerProps = new Properties();
        consumerProps.put("bootstrap.servers", kafkaServer);
        consumerProps.put("group.id", "java-judge-group");
        consumerProps.put("key.deserializer", StringDeserializer.class.getName());
        consumerProps.put("value.deserializer", StringDeserializer.class.getName());

        Properties producerProps = new Properties();
        producerProps.put("bootstrap.servers", kafkaServer);
        producerProps.put("key.serializer", StringSerializer.class.getName());
        producerProps.put("value.serializer", StringSerializer.class.getName());

        KafkaConsumer<String, String> consumer = new KafkaConsumer<>(consumerProps);
        KafkaProducer<String, String> producer = new KafkaProducer<>(producerProps);

        consumer.subscribe(Collections.singletonList("submission-JAVA"));
        System.out.println("[JAVA-JUDGE] Listening for submissions...");

        while (true) {
            ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(100));
            for (ConsumerRecord<String, String> record : records) {
                String json = record.value();
                System.out.println("[RECEIVED] " + json);
                handleSubmission(json, producer, resultEndpoint);
            }
        }
    }

    private static void handleSubmission(String json, KafkaProducer<String, String> producer, String resultEndpoint) throws Exception {
        String code = extractJson(json, "code");
        String submissionId = extractJson(json, "submitId");
        String userId = extractJson(json, "userId");
        String language = extractJson(json, "language");
        String problemId = extractJson(json, "problemId");

        Path workDir = Paths.get("/app/workdir/" + submissionId);
        Files.createDirectories(workDir);

        Path codeFile = workDir.resolve("Main.java");
        String formattedCode = StringEscapeUtils.unescapeJava(code);
        Files.writeString(codeFile, formattedCode);

        Process compile = new ProcessBuilder("javac", codeFile.toString())
                .directory(workDir.toFile())
                .redirectErrorStream(true)
                .start();
        compile.waitFor();

        if (compile.exitValue() != 0) {
            String error = new String(compile.getInputStream().readAllBytes());
            sendResultToServer(submissionId, userId, language, 0, Status.ERROR, error, 0.0, resultEndpoint);
            return;
        }

        String testcaseDir = "/app/testcases/" + problemId + "/testcase/";
        int totalScore = 0;
        int numCases = 0;
        double totalTime = 0.0;

        for (int i = 1; ; i++) {
            Path inputFile = Paths.get(testcaseDir, "input" + i + ".txt");
            Path outputFile = Paths.get(testcaseDir, "output" + i + ".txt");
            if (!Files.exists(inputFile) || !Files.exists(outputFile)) break;

            Path resultFile = workDir.resolve("actual" + i + ".txt");
            long start = System.nanoTime();
            Process run = new ProcessBuilder("timeout", "3s", "java", "Main")
                    .directory(workDir.toFile())
                    .redirectInput(inputFile.toFile())
                    .redirectOutput(resultFile.toFile())
                    .start();
            run.waitFor();
            long end = System.nanoTime();

            double executionTime = (end - start) / 1_000_000_000.0;
            totalTime += executionTime;

            String actual = Files.readString(resultFile).strip();
            String expected = Files.readString(outputFile).strip();

            System.out.println("[TESTCASE " + i + "] Actual: " + actual);
            System.out.println("[TESTCASE " + i + "] Expected: " + expected);

            if (actual.equals(expected)) {
                totalScore += 100;
            }
            numCases++;
        }

        int score = numCases == 0 ? 0 : totalScore / numCases;
        double avgExecutionTime = numCases == 0 ? 0.0 : totalTime / numCases;

        sendResultToServer(submissionId, userId, language, score, Status.PASS, "", avgExecutionTime, resultEndpoint);

        String doneJson = String.format("{\"submitId\":\"%s\",\"problemId\":\"%s\",\"userId\":\"%s\",\"language\":\"%s\"}",
                submissionId, problemId, userId, language);
        producer.send(new ProducerRecord<>("compile-done", doneJson));
        System.out.println("[COMPILE + TESTCASE SUCCESS] sent to topic compile-done");
    }

    private static void sendResultToServer(String submitId, String userId, String language, int score, Status status, String error, double executionTime, String endpoint) throws Exception {
        String apiKey = System.getenv().getOrDefault("RESULT_API_KEY", "WORKER-KEY");

        String body = String.format(
                "{" +
                        "\"submitId\":%s," +
                        "\"score\":%d," +
                        "\"status\":\"%s\"," +
                        "\"executionTime\":%.3f," +
                        "\"errorDetail\":\"%s\"" +
                        "}",
                submitId, score, status, executionTime, error.replace("\"", "'")
        );

        URL url = new URL(endpoint);
        HttpURLConnection con = (HttpURLConnection) url.openConnection();
        con.setRequestMethod("POST");
        con.setRequestProperty("Content-Type", "application/json");
        con.setRequestProperty("X-API-KEY", apiKey);
        con.setDoOutput(true);

        try (OutputStream os = con.getOutputStream()) {
            byte[] input = body.getBytes("utf-8");
            os.write(input, 0, input.length);
        }

        int code = con.getResponseCode();
        System.out.println("[RESULT POST] Response Code: " + code);
    }

    private static String extractJson(String json, String key) {
        String pattern = "\"" + key + "\":(?:\"(.*?)\"|(\\d+))";
        java.util.regex.Matcher matcher = java.util.regex.Pattern.compile(pattern).matcher(json);
        if (matcher.find()) {
            return matcher.group(1) != null ? matcher.group(1) : matcher.group(2);
        }
        return "";
    }
}

enum Status {
    CORRECT,
    WRONG,
    REJECTED,
    PASS,
    ERROR,
    NOT_SUBMITTED
}

File: src/main/resources/docker/cpp/CompileWorker.py
import os
import json
import time
import subprocess
from kafka import KafkaConsumer, KafkaProducer

KAFKA_SERVERS = os.getenv("KAFKA_SERVERS", "localhost:9092")
RESULT_ENDPOINT = os.getenv("RESULT_ENDPOINT", "http://localhost:8082/submit/result")
API_KEY = os.getenv("RESULT_API_KEY", "WORKER-KEY")

consumer = KafkaConsumer(
    "submission-CPP",
    bootstrap_servers=KAFKA_SERVERS,
    value_deserializer=lambda m: json.loads(m.decode("utf-8")),
    auto_offset_reset="earliest",
    group_id="cpp-judge-group"
)

producer = KafkaProducer(
    bootstrap_servers=KAFKA_SERVERS,
    value_serializer=lambda m: json.dumps(m).encode("utf-8")
)

print("[CPP-JUDGE] Listening for submissions...")

def send_result(data):
    import requests
    headers = {
        "Content-Type": "application/json",
        "X-API-KEY": API_KEY
    }
    try:
        res = requests.post(RESULT_ENDPOINT, headers=headers, json=data)
        print(f"[RESULT POST] Status: {res.status_code}")
        if res.status_code >= 400:
            print(f"[SERVER ERROR] {res.text}")
    except Exception as e:
        print(f"[ERROR] Failed to send result: {e}")

for msg in consumer:
    data = msg.value
    print("[RECEIVED]", data)

    submit_id = data["submitId"]
    user_id = data["userId"]
    language = data["language"]
    problem_id = data["problemId"]
    code = data["code"]

    work_dir = f"/app/workdir/{submit_id}"
    os.makedirs(work_dir, exist_ok=True)

    code_path = os.path.join(work_dir, "main.cpp")
    with open(code_path, "w") as f:
        f.write(code)

    executable = os.path.join(work_dir, "a.out")
    compile_cmd = ["g++", code_path, "-o", executable]

    try:
        result = subprocess.run(compile_cmd, capture_output=True, text=True, timeout=10)
        if result.returncode != 0:
            send_result({
                "submitId": submit_id,
                "score": 0,
                "status": "ERROR",
                "executionTime": 0.0,
                "errorDetail": result.stderr
            })
            continue
    except subprocess.TimeoutExpired:
        send_result({
            "submitId": submit_id,
            "score": 0,
            "status": "ERROR",
            "executionTime": 0.0,
            "errorDetail": "Compilation timeout"
        })
        continue

    testcase_dir = f"/app/testcases/{problem_id}/testcase"
    total_score = 0
    total_time = 0.0
    num_cases = 0

    while True:
        i = num_cases + 1
        input_path = os.path.join(testcase_dir, f"input{i}.txt")
        output_path = os.path.join(testcase_dir, f"output{i}.txt")
        if not os.path.exists(input_path) or not os.path.exists(output_path):
            break

        with open(input_path, "r") as fin, open(os.path.join(work_dir, f"actual{i}.txt"), "w") as fout:
            start = time.time()
            try:
                run = subprocess.run([executable], stdin=fin, stdout=fout, timeout=3)
            except subprocess.TimeoutExpired:
                total_time += 3.0
                num_cases += 1
                continue
            end = time.time()
            total_time += (end - start)

        with open(os.path.join(work_dir, f"actual{i}.txt")) as f1, open(output_path) as f2:
            actual = f1.read().strip()
            expected = f2.read().strip()
            if actual == expected:
                total_score += 100
        num_cases += 1

    avg_time = total_time / num_cases if num_cases else 0
    final_score = total_score // num_cases if num_cases else 0

    status = "CORRECT" if score == 100 else "WRONG"

    send_result({
        "submitId": submit_id,
        "score": final_score,
        "status": status,
        "executionTime": round(avg_time, 3),
        "errorDetail": ""
    })

    producer.send("compile-done", {
        "submitId": submit_id,
        "userId": user_id,
        "problemId": problem_id,
        "language": language
    })
    print("[SUCCESS] Result sent and compile-done published.")
